android/java/org/chromium/chrome/browser/crypto_wallet/fragments/ApproveTxBottomSheetDialogFragment.java

```diff
@@ -53,12 +54,13 @@
 public class ApproveTxBottomSheetDialogFragment extends BottomSheetDialogFragment {
     public static final String TAG_FRAGMENT = ApproveTxBottomSheetDialogFragment.class.getName();
 
+    private static ObserverList<ApprovedTxObserver> sApprovedTxObserver = new ObserverList<>();
```

Comment:
I thought about that too, but the problem is the wallet UI is composited by multiple activities (which itself is a problem). The ideal situation will be they are multiple fragments and the notification can be done in a parent activity, but in this case because they are a bunch of activities, there are really three ways to maintain a global state, which all has problems:

 - A singleton
 - A static variable
 - Passing the variable between activities every time calling `startActivityForResult`

If there is another way please suggest?

Reply:
There are examples in chromium sources of using static `ObserverList`. For example in `ApplicationStatus.java`. But how do they prevent of registering early is they call `sApprovedTxObserver = new ObserverList<>()` inside `addApprovedTxObserver` if  `sApprovedTxObserver` is null. So we can do the same. Forgetting to remove could be potentially be a real problem as it could lead to a crash

Reply:
But that's how observers work, we should always and remove them, so it's basically up to a dev to always call remove when it's not needed anymore.

Reply:
> how do they prevent of registering early is they call sApprovedTxObserver = new ObserverList<>() inside addApprovedTxObserver if sApprovedTxObserver is null

Ah ok so what they have there is like a combination of singleton and static. I still think the root problem is we have too many activities which I don't really know why (because I think they can indeed be done with fragments), but I guess it is too late to reform that part so I will proceed with singleton here.